docs/syntax-single-quotes-for-singleton-strings.md

```diff
@@ -0,0 +1,133 @@
+# Use `'` for Singleton Strings
+
+## Summary
+
+Adjust type inference to assume that string literals written using single quotes (eg `'string'`) should always be inferred to have a singleton type, and that strings written using `"double quotes"` should always be typed `string`.
+
+## Motivation
+
+It's quite tricky for us to consistently infer a singleton type in exactly all of the places where it's what the developer intended, and for us to always infer `string` where it's what the developer intended.
+
+Today, our inference strategy is to assume that a string literal should have type `string` if possible, but that type can be constrained to a singleton if necessary.
+
+This strategy is broadly successful, but fails in a few ways.
+
+### Bugs
+
+We have bugs that prevent us from deducing the proper type in certain cases.  For example: https://github.com/luau-lang/luau/issues/1483
+
+```lua
+type Form = "do-not-register" | (() -> ())
+
+local function observer(register: () -> Form) end
+
+observer(function()
+	if math.random() > 0.5 then
+		return "do-not-register" -- TypeError: Type 'string' could not be converted into '"do-not-register" | (() -> ())' Luau(1000)
+	end
+	return function() end
+end)
+```
+
+### Singleton Strings and Generics
+
+The currently implemented inference strategy doesn't always do what the developer intended in the presence of generics.  Whenever a string literal is passed to a generic, it is possible for it to be typed `string`, and so that's what Luau infers.
+
+In the following snippet using the [luaup](https://github.com/jackdotink/luaup) project, a developer clearly wanted `Kind` to be bound to a string singleton and not `string`.  They were instead forced to write a bunch of casts:
+
+```lua
+local function new_token<Kind>(kind: Kind, text: string?): cst.TokenKind<Kind>
+    return { kind = kind, text = text or kind :: any, span = span, trivia = trivias }
```

Comment:
This specific cast is arguably because of the lack of generic bounds. If `TokenKind` defines `text: string`, then `Kind` must be bounded by `string` for `text or kind` to be valid. If anything, this specific example is a little unsound.

Reply:
If you write `type TokenKind<Kind> = { read kind: Kind, read text: string, .. }`, then `new_token(";")` simply infers `TokenKind<";">` because of polarity.